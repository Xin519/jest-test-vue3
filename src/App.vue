<template>
  <section id="app" class="todoapp">
    <header class="header">
      <h1>todos</h1>
      <input
        class="new-todo inputs"
        placeholder="What needs to be done?"
        autocomplete="off"
        autofocus
        v-model="input"
        @keyup.enter="addTodo"
        >
    </header>
    <!-- <p class="inputs">{{ input }}</p> -->
    <section class="main" v-show="count" >
      <input id="toggle-all" class="toggle-all" v-model="allDown" type="checkbox">
      <label for="toggle-all">Mark all as complete</label>
      <ul class="todo-list">
        <li
          v-for="todo in filteredTodos" :key="todo"
          :class="{ editing: todo === editingTodo, completed: todo.completed }"
        >

          <div class="view">
            <input class="toggle" type="checkbox" v-model="todo.completed" >
            <label @dblclick="editTodo(todo)">{{ todo.text }}</label>
            <button class="destroy" @click="remove(todo)"></button>
          </div>
          <input
            class="edit"
            type="text"
            v-editing-focus="todo === editingTodo"
            v-model="todo.text"
            @keyup.enter="doneEdit(todo)"
            @blur="doneEdit(todo)"
            @keyup.esc="cancelEdit(todo)"
            >
        </li>
      </ul>
    </section>
    
    <footer class="footer" v-show="count">
      <span class="todo-count">
        <strong>{{ remainingCount }}</strong> {{ remainingCount > 1 ? 'items' : 'item' }} left
      </span>
      <ul class="filters">
        <li><a href="#/all">All</a></li>
        <li><a href="#/active">Active</a></li>
        <li><a href="#/completed">Completed</a></li>
      </ul>
      <button class="clear-completed"  v-show="count > remainingCount" @click="removeCompleted">
        Clear completed
      </button>
    </footer>

  </section>
 
</template>

<script>
import './assets/index.css'
import useLocalStorage from './utils/useLocalStorage'
import { computed, onMounted, onUnmounted, ref, watchEffect } from 'vue'
const storage = useLocalStorage()

// add
const useAdd = todos => {
  const input = ref('')
  const addTodo = () => {
    const text = input.value && input.value.trim()
    if(text.length === 0) return
    todos.value.unshift({
      text,
      completed: false
    })
    input.value = ''
  }
  return { input, addTodo }
}

// remove
const useRemove = todos => {
  const remove = todo => {
    const inx = todos.value.indexOf(todo)
    todos.value.splice(inx, 1)
  }
  const removeCompleted = () => {
    todos.value = todos.value.filter(todo => !todo.completed)
  }
  return { remove, removeCompleted }
}

// edit
const useEdit = remove => {
  let beforeEditingText = ''
  const editingTodo = ref(null)
  const editTodo = todo => {
    beforeEditingText = todo.text
    editingTodo.value = todo
  }
  const doneEdit = todo => {
    if(!editingTodo.value) return 
    todo.text = todo.text.trim()
    todo.text || remove(todo)
    editingTodo.value = null
  }
  const cancelEdit = todo => {
    editingTodo.value = null
    todo.text = beforeEditingText
  }
  return {
    editingTodo,
    editTodo,
    doneEdit,
    cancelEdit
  }
}

// swith
const useFilter = todos => {
  const allDown = computed({
    get(){
      return !todos.value.filter(todo => !todo.completed).length
    },
    set(value){
      todos.value.forEach(todo => {
        todo.completed = value
      })
    }
  })

  const filter = {
    all: list => list,
    active: list => list.filter(todo => !todo.completed),
    completed: list => list.filter(todo => todo.completed)
  }

  const type = ref('all')
  const filteredTodos = computed(() => filter[type.value](todos.value))
  const remainingCount = computed(() => filter.active(todos.value).length)
  const count = computed(() => todos.value.length)

  const onHashChange = () => {
    const hash = window.location.hash.replace('#/', '')
    if(filter[hash]) {
      type.value = hash
    }else{
      type.value = 'all'
      window.location.hash = ''
    }
  }

  onMounted(() => {
    window.addEventListener('hashchange',onHashChange)
    onHashChange()
  })

  onUnmounted(() => {
    window.removeEventListener('hashchange',onHashChange)
  })

  return { allDown, filteredTodos, remainingCount, count }
}

// storage
const useStorage = () => {
  const KEY = 'TODOKEYS'
  const todos = ref(storage.getItem(KEY) || [])
  watchEffect(() => {
    storage.setItem(KEY, todos.value)
  })
  return todos
}


export default {
  name: 'App',
  setup(){
    const todos = useStorage()
    const { remove, removeCompleted } = useRemove(todos)
    return{
      todos,
      remove,
      removeCompleted,
      ...useAdd(todos),
      ...useEdit(remove),
      ...useFilter(todos)
    }
  },
  directives: {
    editingFocus: (el, binding) => {
      binding.value && el.focus()
    }
  } 
}
</script>

<style>
</style>
