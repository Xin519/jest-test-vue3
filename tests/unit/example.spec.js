import { shallowMount } from '@vue/test-utils'
import App from '@/App.vue'
// const ids = document.getElementById

// describe('HelloWorld.vue', () => {
//   it('renders props.msg when passed', () => {
//     const msg = 'new message'
//     const wrapper = shallowMount(HelloWorld, {
//       props: { msg }
//     })
//     expect(wrapper.text()).toMatch(msg)
//   })
// })

describe('App.vue', () => {
  let wrapper, input, liDom
  // 每个it开始之前执行
  beforeEach(() => {
    wrapper = shallowMount(App)
    input = wrapper.find('.new-todo')
    
  })

  test('input 回车事件有内容发布 addTodo 事件', async () => {
    // 有内容，发布 addTodo 事件 测试预期 向 li 添加数据 内容为 变量 text 数量只有1条
    const text = 'leaf'
    input.setValue(text)
    await input.trigger('keyup.enter')
    liDom = wrapper.find('.todo-list').findAll('li')
    expect(liDom[0].find('label').element.innerHTML).toBe(text)
    expect(liDom.length).toBe(1)
  })

  test('label 双击 修改列表内容', async () => {
    //  测试预期 label 双击 展示文本框 可编辑 编辑后 按esc 为退出 内容不变 按enter修改数据 清空文本按enter 删除 ....
    liDom = wrapper.find('.todo-list').findAll('li')
    const text = liDom[0].find('label').element.innerHTML
    const text1 = 'leaf123'
    liDom[0].find('label').trigger('dblclick')
    liDom[0].find('.edit').setValue(text1)
    await liDom[0].find('.edit').trigger('keyup.esc')
    expect(liDom[0].find('label').element.innerHTML).toBe(text)
    expect(liDom.length).toBe(1)

    liDom[0].find('label').trigger('dblclick')
    liDom[0].find('.edit').setValue(text1)
    await liDom[0].find('.edit').trigger('keyup.enter')
    expect(liDom[0].find('label').element.innerHTML).toBe(text1)
    expect(liDom.length).toBe(1)

    liDom[0].find('label').trigger('dblclick')
    liDom[0].find('.edit').setValue('')
    await liDom[0].find('.edit').trigger('blur')
    liDom = wrapper.find('.todo-list').findAll('li')
    expect(liDom.length).toBe(0)

  })

  test('切换数据状态', async () => {
    input.setValue(text)
    await input.trigger('keyup.enter') // 添加一条数据

    // 测试预期 改变 checkbox 值 切换数据状态 li有无class 'completed'
    liDom = wrapper.find('.todo-list').findAll('li')
    await liDom[0].find('.toggle').setChecked(true)
    expect(liDom[0].classes()).toContain('completed')

    liDom = wrapper.find('.todo-list').findAll('li')
    await liDom[0].find('.toggle').setChecked(false)
    expect(liDom[0].classes()).not.toContain('completed')

  })

  test('点击 x 删除数据', async () => {
    // 测试预期 点击数据后的 x 删除
    liDom = wrapper.find('.todo-list').findAll('li')
    await liDom[0].find('.destroy').trigger('click')
    liDom = wrapper.find('.todo-list').findAll('li')
    expect(liDom.length).toBe(0)
  })
  
  test('点击全选 改变所有数据状态', async () => {
    // 添加数据
    const addText = [
      'leaf1','leaf2','leaf3'
    ]
    await addText.forEach(item => {
      input.setValue(item)
      input.trigger('keyup.enter')
    })

    // 测试预期 点击全选 改变所有数据状态
    liDom = wrapper.find('.todo-list').findAll('li')
    const all = wrapper.find('.toggle-all')
    await all.setChecked(true)
    for(let i = 0; i < liDom.length; i++){
      expect(liDom[i].classes()).toContain('completed')
    }

    await all.setChecked(false)
    for(let r = 0; r < liDom.length; r++){
      expect(liDom[r].classes()).not.toContain('completed')
    }
    
  })

})
