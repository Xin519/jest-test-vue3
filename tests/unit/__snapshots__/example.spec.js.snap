// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App.vue renders props.msg when passed 1`] = `
VueWrapper {
  "__app": Object {
    "_component": Object {
      "__emits": null,
      "__props": Array [
        Object {},
        Array [],
      ],
      "name": "VTU_ROOT",
      "render": [Function],
    },
    "_container": <div
      data-v-app=""
    >
      <section
        class="todoapp"
        id="app"
      >
        <header
          class="header"
        >
          <h1>
            todos
          </h1>
          <input
            autocomplete="off"
            autofocus=""
            class="new-todo"
            placeholder="What needs to be done?"
          />
        </header>
        <section
          class="main"
          style="display: none;"
        >
          <input
            class="toggle-all"
            id="toggle-all"
            type="checkbox"
          />
          <label
            for="toggle-all"
          >
            Mark all as complete
          </label>
          <ul
            class="todo-list"
          >
            
            
          </ul>
        </section>
        <footer
          class="footer"
          style="display: none;"
        >
          <span
            class="todo-count"
          >
            <strong>
              0
            </strong>
             item left 
          </span>
          <ul
            class="filters"
          >
            <li>
              <a
                href="#/all"
              >
                All
              </a>
            </li>
            <li>
              <a
                href="#/active"
              >
                Active
              </a>
            </li>
            <li>
              <a
                href="#/completed"
              >
                Completed
              </a>
            </li>
          </ul>
          <button
            class="clear-completed"
            style="display: none;"
          >
             Clear completed 
          </button>
        </footer>
      </section>
    </div>,
    "_context": Object {
      "app": [Circular],
      "components": Object {
        "transition": Object {
          "name": "transition",
          "props": undefined,
          "render": [Function],
        },
        "transition-group": Object {
          "name": "transition-group",
          "props": undefined,
          "render": [Function],
        },
      },
      "config": Object {
        "errorHandler": undefined,
        "globalProperties": Object {},
        "isCustomElement": [Function],
        "isNativeTag": [Function],
        "optionMergeStrategies": Object {},
        "performance": false,
        "warnHandler": undefined,
      },
      "directives": Object {},
      "mixins": Array [
        Object {
          "__emits": null,
          "__props": Array [],
          "beforeCreate": [Function],
        },
      ],
      "provides": Object {},
      "reload": [Function],
    },
    "_props": null,
    "_uid": 0,
    "component": [Function],
    "config": Object {
      "errorHandler": undefined,
      "globalProperties": Object {},
      "isCustomElement": [Function],
      "isNativeTag": [Function],
      "optionMergeStrategies": Object {},
      "performance": false,
      "warnHandler": undefined,
    },
    "directive": [Function],
    "mixin": [Function],
    "mount": [Function],
    "provide": [Function],
    "unmount": [Function],
    "use": [Function],
    "version": "3.0.10",
  },
  "__setProps": [Function],
  "componentVM": Object {
    "addTodo": [Function],
    "allDown": true,
    "cancelEdit": [Function],
    "count": 0,
    "doneEdit": [Function],
    "editTodo": [Function],
    "editingTodo": null,
    "filteredTodos": Array [],
    "hasOwnProperty": [Function],
    "input": "",
    "remainingCount": 0,
    "remove": [Function],
    "removeCompleted": [Function],
    "todos": Array [],
  },
  "rootVM": Object {},
  "wrapperElement": <section
    class="todoapp"
    id="app"
  >
    <header
      class="header"
    >
      <h1>
        todos
      </h1>
      <input
        autocomplete="off"
        autofocus=""
        class="new-todo"
        placeholder="What needs to be done?"
      />
    </header>
    <section
      class="main"
      style="display: none;"
    >
      <input
        class="toggle-all"
        id="toggle-all"
        type="checkbox"
      />
      <label
        for="toggle-all"
      >
        Mark all as complete
      </label>
      <ul
        class="todo-list"
      >
        
        
      </ul>
    </section>
    <footer
      class="footer"
      style="display: none;"
    >
      <span
        class="todo-count"
      >
        <strong>
          0
        </strong>
         item left 
      </span>
      <ul
        class="filters"
      >
        <li>
          <a
            href="#/all"
          >
            All
          </a>
        </li>
        <li>
          <a
            href="#/active"
          >
            Active
          </a>
        </li>
        <li>
          <a
            href="#/completed"
          >
            Completed
          </a>
        </li>
      </ul>
      <button
        class="clear-completed"
        style="display: none;"
      >
         Clear completed 
      </button>
    </footer>
  </section>,
}
`;

exports[`App.vue 页面是否渲染 1`] = `
VueWrapper {
  "__app": Object {
    "_component": Object {
      "__emits": null,
      "__props": Array [
        Object {},
        Array [],
      ],
      "name": "VTU_ROOT",
      "render": [Function],
    },
    "_container": <div
      data-v-app=""
    >
      <section
        class="todoapp"
        id="app"
      >
        <header
          class="header"
        >
          <h1>
            todos
          </h1>
          <input
            autocomplete="off"
            autofocus=""
            class="new-todo"
            id="inputs"
            placeholder="What needs to be done?"
          />
        </header>
        <section
          class="main"
          style="display: none;"
        >
          <input
            class="toggle-all"
            id="toggle-all"
            type="checkbox"
          />
          <label
            for="toggle-all"
          >
            Mark all as complete
          </label>
          <ul
            class="todo-list"
          >
            
            
          </ul>
        </section>
        <footer
          class="footer"
          style="display: none;"
        >
          <span
            class="todo-count"
          >
            <strong>
              0
            </strong>
             item left 
          </span>
          <ul
            class="filters"
          >
            <li>
              <a
                href="#/all"
              >
                All
              </a>
            </li>
            <li>
              <a
                href="#/active"
              >
                Active
              </a>
            </li>
            <li>
              <a
                href="#/completed"
              >
                Completed
              </a>
            </li>
          </ul>
          <button
            class="clear-completed"
            style="display: none;"
          >
             Clear completed 
          </button>
        </footer>
      </section>
    </div>,
    "_context": Object {
      "app": [Circular],
      "components": Object {
        "transition": Object {
          "name": "transition",
          "props": undefined,
          "render": [Function],
        },
        "transition-group": Object {
          "name": "transition-group",
          "props": undefined,
          "render": [Function],
        },
      },
      "config": Object {
        "errorHandler": undefined,
        "globalProperties": Object {},
        "isCustomElement": [Function],
        "isNativeTag": [Function],
        "optionMergeStrategies": Object {},
        "performance": false,
        "warnHandler": undefined,
      },
      "directives": Object {},
      "mixins": Array [
        Object {
          "__emits": null,
          "__props": Array [],
          "beforeCreate": [Function],
        },
      ],
      "provides": Object {},
      "reload": [Function],
    },
    "_props": null,
    "_uid": 0,
    "component": [Function],
    "config": Object {
      "errorHandler": undefined,
      "globalProperties": Object {},
      "isCustomElement": [Function],
      "isNativeTag": [Function],
      "optionMergeStrategies": Object {},
      "performance": false,
      "warnHandler": undefined,
    },
    "directive": [Function],
    "mixin": [Function],
    "mount": [Function],
    "provide": [Function],
    "unmount": [Function],
    "use": [Function],
    "version": "3.0.10",
  },
  "__setProps": [Function],
  "componentVM": Object {
    "addTodo": [Function],
    "allDown": true,
    "cancelEdit": [Function],
    "count": 0,
    "doneEdit": [Function],
    "editTodo": [Function],
    "editingTodo": null,
    "filteredTodos": Array [],
    "hasOwnProperty": [Function],
    "input": "",
    "remainingCount": 0,
    "remove": [Function],
    "removeCompleted": [Function],
    "todos": Array [],
  },
  "rootVM": Object {},
  "wrapperElement": <section
    class="todoapp"
    id="app"
  >
    <header
      class="header"
    >
      <h1>
        todos
      </h1>
      <input
        autocomplete="off"
        autofocus=""
        class="new-todo"
        id="inputs"
        placeholder="What needs to be done?"
      />
    </header>
    <section
      class="main"
      style="display: none;"
    >
      <input
        class="toggle-all"
        id="toggle-all"
        type="checkbox"
      />
      <label
        for="toggle-all"
      >
        Mark all as complete
      </label>
      <ul
        class="todo-list"
      >
        
        
      </ul>
    </section>
    <footer
      class="footer"
      style="display: none;"
    >
      <span
        class="todo-count"
      >
        <strong>
          0
        </strong>
         item left 
      </span>
      <ul
        class="filters"
      >
        <li>
          <a
            href="#/all"
          >
            All
          </a>
        </li>
        <li>
          <a
            href="#/active"
          >
            Active
          </a>
        </li>
        <li>
          <a
            href="#/completed"
          >
            Completed
          </a>
        </li>
      </ul>
      <button
        class="clear-completed"
        style="display: none;"
      >
         Clear completed 
      </button>
    </footer>
  </section>,
}
`;
